sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -166,6 +197,25 @@ class FileStreamSource(
         // implies "sourceOptions.latestFirst = true" which we want to refresh the list per batch
         (newFiles.take(files.maxFiles()), null)
 
+      case files: ReadMaxBytes if !sourceOptions.latestFirst =>
+        // we can cache and reuse remaining fetched list of files in further batches
+        val (FilesSplit(bFiles, _), FilesSplit(usFiles, rSize)) =
+          takeFilesUntilMax(newFiles, files.maxBytes())
+        if (rSize < (files.maxBytes() * DISCARD_UNSEEN_FILES_RATIO).toLong) {
```

Comment:
Hmm, if `BigInt` is used to avoid overflow, does it make sense to cast it to `Long` to compare here? `rSize` might be larger than long range.

Reply:
My reasoning was that `files.maxBytes()` is Long and `0 < DISCARD_UNSEEN_FILES_RATIO < 1` so the result must not exceed `Long.MaxValue`.

Reply:
I think `rSize` is `BigInt`? If it is larger than `Long.MaxValue`, `BigInt.toLong` will return a value which is smaller than actual value? So the condition is wrongly evaluated to true.

Reply:
`rSize` is BigInt yes
`(files.maxBytes() * DISCARD_UNSEEN_FILES_RATIO).toLong` is Long
So it `rSize` > `Long.MaxValue` then it also means `rSize` > `(files.maxBytes() * DISCARD_UNSEEN_FILES_RATIO).toLong` and our expression will be evaluated to `False` which is the behavior that we want as I understand.
Am I missing something?

Reply:
I mean, if you do `rSize.toLong` (https://github.com/apache/spark/pull/44636#discussion_r1470701871), then

```
if (rSize.toLong < (files.maxBytes() * DISCARD_UNSEEN_FILES_RATIO).toLong) {
  ...
}
```

`toLong` will return truncated value. It may make this condition as true.

Reply:
Got it, I [commented](https://github.com/apache/spark/pull/44636#discussion_r1472331686) on it above. As I understand `DISCARD_UNSEEN_FILES_RATIO` is not supposed to be bigger then 1 so we can't have a truncation issue. Let me know if we need adjustments.

Reply:
I don't mean there will be truncation on `files.maxBytes() * DISCARD_UNSEEN_FILES_RATIO).toLong`, but `rSize.toLong`.

Reply:
To be clear, If `rSize` is a value larger than `Lonx.MaxValue`, once you did `rSize.toLong`, it may possibly cast to a long value truncated to 1, for example. So the condition `rSize.toLong < (files.maxBytes() * DISCARD_UNSEEN_FILES_RATIO).toLong` will wrongly evaluate to `true`, although `rSize` is actually larger than `files.maxBytes() * DISCARD_UNSEEN_FILES_RATIO).toLong`.

To compare two values in different ranges, it seems more reasonable to compare them in wider range, instead of narrow one (as you will truncate the value from wider range).

Reply:
@viirya 
I've converted BigInt to Double and now it is a comparison of two Doubles
let me know if that is acceptable, we could also do:
```
if (rSize < BigDecimal(files.maxBytes() * DISCARD_UNSEEN_INPUT_RATIO)) {
```
BigInt on the left would be converted to BigDecimal implicitly